core/services/postgres/lease_lock.go

```diff
@@ -0,0 +1,188 @@
+	"github.com/smartcontractkit/sqlx"
+)
+
+// Lessee - /lɛˈsiː/
+// noun
+// a person who holds the lease of a property; a tenant
+
+// The lease lock (or lessee) handles taking an exclusive lease on database access. This is voluntary and not enforced by any database primitives, but rather voluntarily respected by other instances of the Chainlink application.
+//
+// Chainlink is designed to run as a single instance. Running multiple instances of Chainlink on a single database at the same time is not supported and likely to lead to strange errors and possibly even data integrity failures.
+//
+// With that being said, a very common use case is to run multiple Chainlink instances in failover mode. The first instance will take some kind of lock on the database and subsequent instances will wait trying to take this lock in case the first instance disappears or dies.
+//
+// Traditionally Chainlink has used an advisory lock to manage this. However, advisory locks come with several problems, notably:
+// - Postgres does not really like it when you hold locks open for a very long time (hours/days). It hampers certain internal cleanup tasks and is explicitly discouraged by the postgres maintainers
+// - The advisory lock can silently disappear on postgres upgrade
+// - Advisory locks do not play nicely with pooling tools such as pgbouncer
+// - If the application crashes, the advisory lock can be left hanging around for a while (sometimes hours) and can require manual intervention to remove it
+//
+// For this reason, we now use a database lessee instead, which works as such:
+// - Have one row in a database which is updated periodically with the client ID
+// - CL node A will run a background process on start that updates this e.g. once per second
+// - CL node B will spinlock, checking periodically to see if the update got too old. If it goes more than, say, 5s without updating, it assumes that node A is dead and takes over. Now CL node B is the owner of the row and it updates this every second
+// - If CL node A comes back somehow, it will go to take out a lease and realise that the database has been leased to another process, so it will panic and quit immediately
+type Lessee interface {
+	TakeAndHold() error
+	Release()
+	ClientID() uuid.UUID
+}
+
+var _ Lessee = &lessee{}
+
+type lessee struct {
+	// TODO: Use a "master" application parent ctx that is cancelled on stop
+	id              uuid.UUID
+	db              *sqlx.DB
+	refreshInterval time.Duration
+	leaseDuration   time.Duration
+	logger          logger.Logger
+
+	chStop chan struct{}
+	wg     sync.WaitGroup
+}
+
+// NewLessee creates a "lessee" - an entity that tries to take an exclusive lease on the database
+func NewLessee(db *sqlx.DB, appID uuid.UUID, lggr logger.Logger, refreshInterval, leaseDuration time.Duration) Lessee {
+	return &lessee{appID, db, refreshInterval, leaseDuration, lggr.Named("Lessee").With("appID", appID), make(chan struct{}), sync.WaitGroup{}}
```

Comment:
It might be good to sanity check for reasonable durations here. Maybe `refreshInterval` must be above some threshold, and `leaseDuration` must be > some multiple of `refreshInterval`?

Do you think we'll want/need to tune them for different types of deployments? e.g. calibrate based on DB latency if it's not local. Or could it make sense to derive them directly from our default DB call timeout in some way?

Reply:
We only set it up in once place, so I think it's ok, but I can enforce a basic sanity check that refresh interval must be at most the same as lease duration